### PR TITLE
Fix bug filtering collinear points on vertical lines

### LINKED ISSUE #81076: geotile_grid aggregation on geo_shape produces incorrect results for polygons in certain corner cases
**Elasticsearch version** (`bin/elasticsearch --version`): 7.13.2-SNAPSHOT

**Description of the problem including expected versus actual behavior**:

Certain geo_shapes cause the geotile_grid aggregation to produce incorrect results.  

**Steps to reproduce**:

1. Create the index
```
PUT /test-data
{
  "settings": {
    "number_of_shards": 1
  },
  "mappings": {
    "properties": {
      "shape": { "type": "geo_shape" },
      "time": { "type": "date" }
    }
  }
}
```

2. Put a rectangle
```
POST /test-data/_doc
{
    "shape" : {
        "type" : "polygon",
        "coordinates" : [
            [ [180.0, 20.54099347935443], [180.00, 56.22356828010862], [173.32664998748726, 56.22356828010862], [173.32664998748726, 20.54099347935443], [180.0, 20.54099347935443] ]
        ]
    },
    "time": "2021-11-26T00:00:00Z"
}
```

3. Do an aggregation and observe correct results.
```
POST test-data/_search?size=0
{
  "query": {
    "match_all": {}
    
  },
  "aggs": {
    "grids": {"geotile_grid": {"field": "shape", "precision": 4}}
  }
}

"buckets" : [
        {
          "key" : "4/15/7",
          "doc_count" : 1
        },
        {
          "key" : "4/15/6",
          "doc_count" : 1
        },
        {
          "key" : "4/15/5",
          "doc_count" : 1
        },
        {
          "key" : "4/15/4",
          "doc_count" : 1
        }
  ]
```

4. Put a more complex polygon that is contained within the same area as the previous shape

```
POST /test-data/_doc
{
    "shape" : {
        "type" : "polygon",
        "coordinates" : [
    [
      [
        180,
        38.382280879731525
      ],
      [
        180,
        39.20402540758703
      ],
      [
        180,
        40.20983928828463
      ],
      [
        180,
        41.5905083136395
      ],
      [
        180,
        43.79069222748844
      ],
      [
        180,
        48.05737800943521
      ],
      [
        180,
        56.22356828010862
      ],
      [
        180,
        53.61517437197104
      ],
      [
        178.17058562712035,
        47.44994633100819
      ],
      [
        177.161329896696,
        44.42163705320016
      ],
      [
        176.63491363226575,
        42.6928738950413
      ],
      [
        176.29683812379142,
        41.52175761146341
      ],
      [
        176.0458600769842,
        40.61841573973024
      ],
      [
        175.83768485617443,
        39.84528044056929
      ],
      [
        175.64814528805073,
        39.12101916544052
      ],
      [
        175.4602369741285,
        38.382280879731525
      ],
      [
        175.2576316219736,
        37.56053635187602
      ],
      [
        175.01887227795362,
        36.55472247117842
      ],
      [
        174.7078894332312,
        35.174053445823546
      ],
      [
        174.25388021927512,
        32.97386953197461
      ],
      [
        173.53771573775157,
        28.707183750027852
      ],
      [
        173.32664998748726,
        20.54099347935443
      ],
      [
        177.52259404863642,
        23.14938738749197
      ],
      [
        180,
        29.314615428454847
      ],
      [
        180,
        32.34292470626289
      ],
      [
        180,
        34.07168786442174
      ],
      [
        180,
        35.24280414799964
      ],
      [
        180,
        36.14614601973281
      ],
      [
        180,
        36.91928131889376
      ],
      [
        180,
        37.64354259402253
      ],
      [
        180,
        38.382280879731525
      ]
    ]
  ]
    },
    "time": "2021-11-26T00:00:00Z"
}
```

5. Do another aggregation
 
```
POST test-data/_search?size=0
{
  "query": {
    "match_all": {}
    
  },
  "aggs": {
    "grids": {"geotile_grid": {"field": "shape", "precision": 4}}
  }
}
```

6. Observe that you get buckets way outside of the expected area.  Picture from Kibana makes it very clear.

image

### PR DESCRIPTION
**Pull request by iverase**
In #59501 we added a process to remove collinear points that exist in vertical lines for polygons as they were making the polygon decomposition logic to fail. Unfortunately this logic introduced a bug while filtering the points that is fixed here.

fixes #81076

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/geo/GeoPolygonDecomposer.java [hunk 1]
```diff
@@ -111,19 +102,30 @@ private static LinearRing filterRing(LinearRing linearRing) {
...
 
+    private static boolean skipPoint(LinearRing linearRing, int i) {
+        if (linearRing.getLon(i - 1) == linearRing.getLon(i)) {
+            if (linearRing.getLat(i - 1) == linearRing.getLat(i)) {
+                // same point
+                return true;
+            }
+            if (linearRing.getLon(i - 1) == linearRing.getLon(i + 1)
+                && linearRing.getLat(i - 1) > linearRing.getLat(i) != linearRing.getLat(i + 1) > linearRing.getLat(i)) {
+                // coplanar
```
Comment (imotov) [inline server/src/main/java/org/elasticsearch/common/geo/GeoPolygonDecomposer.java:122]: I am a bit confused about terminology and the logic here. We have only 2 coordinates, so theoretically all our points are coplanar. Do you mean collinear here? Based on this logic we only remove points if we go down and then up within the same longitude. What's the reason for not removing colinear points in between? Basically, in the example below: `POLYGON ((0 0, 0 10, 10 10, 10 5, 10 -10, 10 0, 0 0))` the current logic removes point B, but keep point A intact. 

image

**Review by imotov — COMMENTED** [review]
@iverase, ping me when you have a chance. I think I am missing something here.

File: server/src/main/java/org/elasticsearch/common/geo/GeoPolygonDecomposer.java (hunk 1 above)
Comment (iverase) [reply server/src/main/java/org/elasticsearch/common/geo/GeoPolygonDecomposer.java:122]: doh! Yes, I meant collinear points

Comment (iverase) [reply server/src/main/java/org/elasticsearch/common/geo/GeoPolygonDecomposer.java:122]: I had a look and the current logic that what I expect that is to remove point A:

original:

image

processed:

image

**Review by imotov — APPROVED** [review]
LGTM. Please add comments around this check so future me will not get confused again :)

**Pull request merged**